feat: initial commit

satazor opened the pull request.

andrefgneves commented on README.md:
```diff
@@ -1 +1,285 @@
+
+### ChainedMap
+
+A ChainedMap operates similar to a JavaScript Map, with some conveniences for chaining and generating data structures. It extends [Chainable](#chainable) which means that all its methods are also available.
```
~similar~ similarly

satazor replied:
Will be fixed in the next commit.

andrefgneves commented on README.md:
```diff
@@ -1 +1,285 @@
+<details>
+  <summary><b>Diferences from webpack-chain:</b></summary>
+  <ul>
+    <li>Removed <code>order()</code> because its an internal method, reducing the number of conflicts in case you create a class that inherits from <code>ChainedMap</code></li>
```
~its~ it's

satazor replied:
Will be fixed in the next commit.

andrefgneves commented on README.md:
```diff
@@ -1 +1,285 @@
+
+#### .set(key, value)
+
+Set the value the entry corresponding to `key`.
```
_of_ the entry

satazor replied:
Will be fixed in the next commit.

andrefgneves commented on README.md:
```diff
@@ -1 +1,285 @@
+
+#### .delete(key)
+
+Remove entry corresponding to `key`.
```
_the_ entry

satazor replied:
Will be fixed in the next commit.

andrefgneves commented on README.md:
```diff
@@ -1 +1,285 @@
+
+Returns an array of the keys stored in the Map.
+
+The order will having into consideration entries that called `.before(name)` or `.after(name)`, see [OrderableChainedMap](#orderable-chained-map).
```
_The result will have into consideration the calls to `.before(name)` or `.after(name)`. See [OrderableChainedMap](#orderable-chained-map)._

satazor replied:
Will be fixed in the next commit.

andrefgneves commented on README.md:
```diff
@@ -1 +1,285 @@
+
+Returns an array of the values stored in the Map.
+
+The order will having into consideration entries that called `.before(name)` or `.after(name)`, see [OrderableChainedMap](#orderable-chained-map).
```
_The result will have into consideration the calls to `.before(name)` or `.after(name)`. See [OrderableChainedMap](#orderable-chained-map)._

andrefgneves commented on README.md:
```diff
@@ -1 +1,285 @@
+
+Returns an object of all the entries stored in the Map.
+
+The order will having into consideration entries that called `.before(name)` or `.after(name)`, see [OrderableChainedMap](#orderable-chained-map).
```
_The result will have into consideration the calls to `.before(name)` or `.after(name)`. See [OrderableChainedMap](#orderable-chained-map)._

satazor replied:
Will be fixed in the next commit.

andrefgneves commented on README.md:
```diff
@@ -1 +1,285 @@
+
+Execute `fn` once per each key/value pair in the Map.
+
+The order will having into consideration entries that called `.before(name)` or `.after(name)`, see [OrderableChainedMap](#orderable-chained-map).
```
_The result will have into consideration the calls to `.before(name)` or `.after(name)`. See [OrderableChainedMap](#orderable-chained-map)._

satazor replied:
Will be fixed in the next commit.

andrefgneves commented on README.md:
```diff
@@ -1 +1,285 @@
+
+#### .entries()
+
+Returns an object of all the entries stored in the Map.
```
Should we add a note that this does not mimic the native `entries` API, which returns an array?

acostalima replied:
The following:

"Returns an object containing all the entries stored in the Map"

may sound a little better?

acostalima replied:
@andrefgneves it is already pointed out it returns an object, so I'm not sure that would be necessary.

satazor replied:
Many other maps do not mimic the native API, I don't think its necessary to say that, at least for now.

andrefgneves commented on README.md:
```diff
@@ -1 +1,285 @@
+
+The order will having into consideration entries that called `.before(name)` or `.after(name)`, see [OrderableChainedMap](#orderable-chained-map).
+
+#### .merge(ob, [omit])
```
What do you think of changing `omit` to `exclude` or `skip`? The point being that those properties will still be in the resulting object if they were already in the original Map.

satazor replied:
I want to be as close as possible to `webpack-chain`, will leave it like this for now.

andrefgneves commented on README.md:
```diff
@@ -1 +1,285 @@
+const chainedMap = new ChainedMap();
+
+chainedMap.extend(['foo', 'bar']);
+chainedMap.foo('foz');
```
Should we provide a second argument here to illustrate that these are only setters? Calling these like this would return the ChainedMap, and people might expect these shorthands to also be getters.

satazor replied:
Will add `.get` examples illustrating that they are setters

acostalima commented:
LGMT 😄 👍 
What do you think of `chainables` instead of `base-chain`?

andrefgneves commented on README.md:
```diff
@@ -1 +1,285 @@
+
+### ChainedSet
+
+A ChainedMap operates similar to a JavaScript Set, with some conveniences for chaining and generating data structures. It extends [Chainable](#chainable) which means that all its methods are also available.
```
~similar~ similarly

andrefgneves commented on README.md:
```diff
@@ -1 +1,286 @@
+    <li>Fixed minor bugs that caused <code>.keys()</code>, <code>.values()</code> and <code>.entries()</code> to not respect the order specified with <code>.before(key)</code> and <code>.after(key)</code>
+  </ul>
+</details>
+
```
I think some example usages would be helpful here. It's not entirely obvious whether you can chain a `.before/.after` block and call `.end` or you need a reference to the value and just "move" it with `.before/.after`.

satazor replied:
Agree, I will actually add a `Examples` in each section.

andregoncalvesdev commented on README.md:
```diff
@@ -1 +1,286 @@
+This is the base class that others inherit from.
+
+<details>
+  <summary><b>Diferences from webpack-chain:</b></summary>
```
Differences

satazor replied:
Will be fixed in the next commit.

andregoncalvesdev commented on README.md:
```diff
@@ -1 +1,286 @@
+
+### ChainedSet
+
+A ChainedMap operates similar to a JavaScript Set, with some conveniences for chaining and generating data structures. It extends [Chainable](#chainable) which means that all its methods are also available.
```
A ChainedSet

satazor replied:
Will be fixed in next commit.

satazor commented:
I've updated the MR with everyone's feedback. I've also added examples to every API section.

I would like your review on the `OrderableChainableMap`, I've changed it to be more flexible.

satazor commented on README.md:
```diff
@@ -1 +1,401 @@
+        .set('setup-foo', '/path/to/setup/foo.js')
+        .set('setup-bar', '/path/to/setup/bar.js')
+        .set('setup-baz', '/path/to/setup/baz.js')
+        .order('setup-bar', ({ before }) => before('setup-foo'))
```
I ended up with this signature because it's more semantic, but here's a few alternatives:

`.orderAfter('setup-bar', 'setup-foo')` / `.orderBefore('setup-bar', 'setup-foo')` - not every obvious which one is after what
`order('setup-bar', 'before', 'setup-foo')`
`order('setup-bar', { relativeTo: 'setup-foo', position: 'after' })`

I prefer the current approach because it's more semantic and descriptive. It also goes more inline with `after` and `before` methods added to child Chainables, but I would like your opinion.

andrefgneves replied:
I like this approach too 👍

What do you think of renaming it to `reorder` or `move`?

satazor replied:
I like both actually! Lets wait for @acostalima and @andregoncalvesdev feedback on their preference.

satazor replied:
I've talked with @andregoncalvesdev and we settled on `move`.

satazor commented:
@acostalima regarding the naming, I like `chainables`. The reason I named `base-chain` is that later on we can create `jest-chain`, etc just like `webpack-chain`.

Though, I don't know if that will ever happen.

satazor commented:
I guess this is good to be merged. The only pending discussion is the name. Could you guys weight in?

andrefgneves commented:
I'm ok with `js-base-chain`.

satazor commented:
Ok, I will merge this for now. @andregoncalvesdev could you give your opinion about the name later?

satazor commented:
Omg, I tried to publish the package as `base-chain` but I can't because there's a similar package called `basechain` and npm prevents developers from publishing similar packages now:

> Package name too similar to existing packages; try renaming your package to '@satazor/base-chain' and publishing with 'npm publish --access=public' instead : base-chain

Alternatives:

- core-chain
- root-chain
- low-chain
- basis-chain
- base-chainable
- chainables

Any other suggestions?

andregoncalvesdev commented:
I like chainables and core-chain. Other alternatives: refined-chain, simple-chain,
awesome-chain.

On Wed, Mar 28, 2018 at 12:18 AM, André Cruz <notifications@github.com>
wrote:

> Omg, I tried to publish the package as base-chain but I can't because
> there's a similar package called basechain and npm prevents developers
> from publishing similar packages now:
>
> Package name too similar to existing packages; try renaming your package
> to '' and publishing with 'npm publish --access=public' instead :
> base-chain
>
> Alternatives:
>
>    - base-chainable
>    - chainables
>
> Any other suggestions?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/moxystudio/js-base-chain/pull/1#issuecomment-376706680>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI13ieRq_1MIWvw9ayzd2tj3d04ncOaDks5tisi4gaJpZM4S5u9H>
> .
>



-- 
André Gonçalves,
Developer <https://github.com/andregoncalvesdev>

andrefgneves commented:
* `chainable-config`
* `chained-config`
* `config-chainables`

¯\\_(ツ)_/¯

satazor commented:
@andrefgneves we are more headed towards low-chain or core-chain.

Thoughts?

satazor commented:
I also like chained-config! We must change several references from data to config but it fits well!

satazor commented:
I’m so undecided, could we do a voting? :p

andrefgneves commented:
1️⃣ `low-chain`
2️⃣ `core-chain`
3️⃣ `chained-config`
4️⃣ `chainables`

Let me know if you want me to add other options.

andrefgneves commented:
Oh, crap. I thought we could add any reaction to a comment :/

andrefgneves commented:
3️⃣

andrefgneves commented:
1️⃣ `low-chain`

andrefgneves commented:
2️⃣ `core-chain`

andrefgneves commented:
3️⃣ `chained-config`

andrefgneves commented:
4️⃣ `chainables`

satazor commented:
chained-config it is
